File: Python/textbook_track/chapter11/ba11h.py
# Compute the Size of a Spectral Dictionary
import functools as ft
import sys

from textbook_track.resources.utils import read_amino_acid_mass_table


def convert_to_intlist(line):
    return [int(elem) for elem in line.split()]


def calc_spectral_dictionary_size(spectrum, mass_table, threshold, max_score):
    amino_acid_masses = tuple(mass_table.values())
    m = len(spectrum)

    @ft.lru_cache(maxsize=None)
    def solve(ix, jy):
        if ix == 0 and jy == 0:
            return 1
        if ix <= 0:
            return 0
        return sum(solve(ix-mass, jy-spectrum[ix-1]) for mass in amino_acid_masses)

    return sum(solve(m, score) for score in range(threshold, max_score+1))


def main():
    reader = sys.stdin
    spectrum = convert_to_intlist(next(reader))
    threshold = int(next(reader))
    max_score = int(next(reader))
    mass_table = read_amino_acid_mass_table()
    result = calc_spectral_dictionary_size(spectrum, mass_table, threshold, max_score)
    print(result)


if __name__ == '__main__':
    main()

File: Scala/src/main/scala/TextbookTrack/Chapter11/BA11H.scala
package TextbookTrack.Chapter11

object BA11H {
  import TextbookTrack.Utils.IntegerMassTable
  import scala.collection.mutable.{Map => MutableMap}

  private def convertToIntVector(line: String): Vector[Int] = line.split(" ").map(_.toInt).toVector

  def calcSpectralDictionarySize(spectrum: Vector[Int],
                                 massTable: Map[Char, Int],
                                 threshold: Int,
                                 maxScore: Int): Int = {
    val aminoAcidMasses: List[Int] = massTable.valuesIterator.toList
    val m: Int = spectrum.length
    val cache: MutableMap[(Int, Int), Int] = MutableMap()

    def solve(ix: Int, jy: Int): Int = {
      if (ix == 0 && jy == 0) 1
      else if (ix <= 0) 0
      else {
        def score: Int = aminoAcidMasses.foldLeft(0)((acc, mass) => acc + solve(ix - mass, jy - spectrum(ix - 1)))
        cache.getOrElseUpdate((ix, jy), score)
      }
    }

    (threshold to maxScore).foldLeft(0)((acc, t) => acc + solve(m, t))
  }

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val spectrum: Vector[Int] = convertToIntVector(reader.next())
    val threshold: Int = reader.next().toInt
    val maxScore: Int = reader.next().toInt
    val result: Int = calcSpectralDictionarySize(spectrum, IntegerMassTable, threshold, maxScore)
    println(result)
  }
}
